File: otto-mobile/otto-mobile/ChatEngagementThrottle.swift
import Foundation

enum ChatEngagementAlertOutcome: Equatable {
    case suppress
    case fullAlert
    case silentBannerUpdate(suppressedCount: Int)
}

@MainActor
enum ChatEngagementThrottle {
    private static let cooldownTiers: [TimeInterval] = [15, 30, 45, 60]
    private static let idleDecayInterval: TimeInterval = 60

    private struct ConversationNotificationState {
        var lastAlertAt: Date?
        var lastActivityAt: Date?
        var cooldownUntil: Date?
        var suppressedCount: Int = 0
        var tierIndex: Int = 0
    }

    private static var stateByKey: [String: ConversationNotificationState] = [:]

    static func squadConversationKey(circleId: String) -> String {
        let trimmed = circleId.trimmingCharacters(in: .whitespacesAndNewlines)
        return "squad:\(trimmed)"
    }

    static func directConversationKey(conversationId: String) -> String {
        let trimmed = conversationId.trimmingCharacters(in: .whitespacesAndNewlines)
        return "dm:\(trimmed)"
    }

    static func foregroundNotificationIdentifier(forConversationKey key: String) -> String {
        "otto-chat-foreground-\(key.replacingOccurrences(of: ":", with: "-"))"
    }

    static func evaluateSquadMessage(
        circleId: String,
        pushType: String = "circle.chat.new_message",
        focusedChatCircleId: String?,
        now: Date = Date()
    ) -> ChatEngagementAlertOutcome {
        let trimmedCircle = circleId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedCircle.isEmpty else { return .suppress }

        if focusedChatCircleId == trimmedCircle {
            return .suppress
        }
        return evaluateCooldown(forKey: squadConversationKey(circleId: trimmedCircle), now: now)
    }

    static func evaluateDirectMessage(
        conversationId: String,
        focusedConversationId: String?,
        now: Date = Date()
    ) -> ChatEngagementAlertOutcome {
        let trimmedConversation = conversationId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedConversation.isEmpty else { return .suppress }

        if focusedConversationId == trimmedConversation {
            return .suppress
        }
        return evaluateCooldown(forKey: directConversationKey(conversationId: trimmedConversation), now: now)
    }

    static func evaluatePushPresentation(
        type: String,
        circleId: String?,
        conversationId: String?,
        now: Date = Date()
    ) -> ChatEngagementAlertOutcome {
        switch type {
        case "circle.chat.mention", "circle.chat.reply", "circle.chat.new_message":
            guard let circleId else { return .fullAlert }
            return evaluateSquadMessage(
                circleId: circleId,
                pushType: type,
                focusedChatCircleId: PushFocusBridge.activeChatCircleId,
                now: now
            )
        case "direct.message":
            guard let conversationId else { return .fullAlert }
            return evaluateDirectMessage(
                conversationId: conversationId,
                focusedConversationId: PushFocusBridge.activeDirectConversationId,
                now: now
            )
        default:
            return .fullAlert
        }
    }

    static func recordFullAlert(
        forConversationKey key: String,
        pushType: String = "circle.chat.new_message",
        now: Date = Date()
    ) {
        var state = stateByKey[key] ?? ConversationNotificationState()
        applyIdleDecay(state: &state, now: now)
        let tiered = usesTieredCooldown(pushType: pushType)
        let duration = cooldownDuration(tierIndex: state.tierIndex, tiered: tiered)
        state.lastAlertAt = now
        state.lastActivityAt = now
        state.cooldownUntil = now.addingTimeInterval(duration)
        if tiered, state.suppressedCount > 0 {
            state.tierIndex = min(state.tierIndex + 1, cooldownTiers.count - 1)
        }
        state.suppressedCount = 0
        stateByKey[key] = state
    }

    static func recordSilentUpdate(
        forConversationKey key: String,
        pushType: String = "circle.chat.new_message",
        now: Date = Date()
    ) {
        var state = stateByKey[key] ?? ConversationNotificationState()
        applyIdleDecay(state: &state, now: now)
        state.suppressedCount += 1
        state.lastActivityAt = now
        let tiered = usesTieredCooldown(pushType: pushType)
        if state.cooldownUntil == nil {
            state.cooldownUntil = now.addingTimeInterval(cooldownDuration(tierIndex: state.tierIndex, tiered: tiered))
        }
        stateByKey[key] = state
    }

    static func bannerBody(base: String, suppressedCount: Int) -> String {
        guard suppressedCount > 0 else { return base }
        return "\(base) (+\(suppressedCount) more)"
    }

    /// Clears in-memory throttle state (unit tests).
    static func resetForTesting() {
        stateByKey.removeAll()
    }

    /// Current tier index for a conversation (unit tests).
    static func tierIndexForTesting(conversationKey: String) -> Int {
        stateByKey[conversationKey]?.tierIndex ?? 0
    }

    /// Cooldown deadline for a conversation (unit tests).
    static func cooldownUntilForTesting(conversationKey: String) -> Date? {
        stateByKey[conversationKey]?.cooldownUntil
    }

    private static func usesTieredCooldown(pushType: String) -> Bool {
        pushType == "circle.chat.new_message"
    }

    private static func cooldownDuration(tierIndex: Int, tiered: Bool) -> TimeInterval {
        if tiered {
            return cooldownTiers[min(max(tierIndex, 0), cooldownTiers.count - 1)]
        }
        return cooldownTiers[0]
    }

    private static func applyIdleDecay(state: inout ConversationNotificationState, now: Date) {
        guard state.tierIndex > 0, let lastActivity = state.lastActivityAt else { return }
        let idleSeconds = now.timeIntervalSince(lastActivity)
        let stepsDown = min(state.tierIndex, Int(idleSeconds / idleDecayInterval))
        guard stepsDown > 0 else { return }
        state.tierIndex -= stepsDown
        state.suppressedCount = 0
    }

    private static func evaluateCooldown(forKey key: String, now: Date) -> ChatEngagementAlertOutcome {
        var state = stateByKey[key] ?? ConversationNotificationState()
        applyIdleDecay(state: &state, now: now)
        state.lastActivityAt = now
        stateByKey[key] = state

        if let cooldownUntil = state.cooldownUntil, now < cooldownUntil {
            return .silentBannerUpdate(suppressedCount: state.suppressedCount + 1)
        }
        return .fullAlert
    }
}

File: otto-android/app/src/main/java/to/ottomot/driftd/core/notify/ChatEngagementThrottle.kt
package to.ottomot.driftd.core.notify

import java.util.concurrent.ConcurrentHashMap

enum class ChatEngagementAlertOutcome {
    Suppress,
    FullAlert,
    SilentBannerUpdate,
}

/**
 * Per-conversation foreground alert cooldown with tier ladder [15, 30, 45, 60]s.
 * Escalates after burst silent updates; decays one tier per 60s idle.
 * Mentions, replies, and DMs always use 15s and never change tier.
 */
object ChatEngagementThrottle {
    private val cooldownTiersMs = longArrayOf(15_000L, 30_000L, 45_000L, 60_000L)
    private const val IDLE_DECAY_MS = 60_000L

    private data class ConversationNotificationState(
        var lastAlertAtMs: Long? = null,
        var lastActivityAtMs: Long? = null,
        var cooldownUntilMs: Long? = null,
        var suppressedCount: Int = 0,
        var tierIndex: Int = 0,
    )

    private val stateByKey = ConcurrentHashMap<String, ConversationNotificationState>()

    fun squadConversationKey(circleId: String): String {
        val trimmed = circleId.trim()
        return "squad:$trimmed"
    }

    fun directConversationKey(conversationId: String): String {
        val trimmed = conversationId.trim()
        return "dm:$trimmed"
    }

    fun foregroundNotificationId(conversationKey: String): Int =
        "otto.chat.foreground.$conversationKey".hashCode()

    fun evaluateSquadMessage(
        circleId: String,
        pushType: String = "circle.chat.new_message",
        focusedChatCircleId: String?,
        isMuted: Boolean,
        nowMs: Long = System.currentTimeMillis(),
    ): Pair<ChatEngagementAlertOutcome, Int> {
        val trimmedCircle = circleId.trim()
        if (trimmedCircle.isEmpty()) return ChatEngagementAlertOutcome.Suppress to 0
        if (focusedChatCircleId != null &&
            focusedChatCircleId.trim().equals(trimmedCircle, ignoreCase = true)
        ) {
            return ChatEngagementAlertOutcome.Suppress to 0
        }
        if (isMuted) return ChatEngagementAlertOutcome.Suppress to 0
        return evaluateCooldown(squadConversationKey(trimmedCircle), nowMs)
    }

    fun evaluateDirectMessage(
        conversationId: String,
        focusedConversationId: String?,
        nowMs: Long = System.currentTimeMillis(),
    ): Pair<ChatEngagementAlertOutcome, Int> {
        val trimmedConversation = conversationId.trim()
        if (trimmedConversation.isEmpty()) return ChatEngagementAlertOutcome.Suppress to 0
        if (focusedConversationId != null &&
            focusedConversationId.trim().equals(trimmedConversation, ignoreCase = true)
        ) {
            return ChatEngagementAlertOutcome.Suppress to 0
        }
        return evaluateCooldown(directConversationKey(trimmedConversation), nowMs)
    }

    fun recordFullAlert(
        conversationKey: String,
        pushType: String = "circle.chat.new_message",
        nowMs: Long = System.currentTimeMillis(),
    ) {
        val state = stateByKey.getOrPut(conversationKey) { ConversationNotificationState() }
        applyIdleDecay(state, nowMs)
        val tiered = usesTieredCooldown(pushType)
        val durationMs = cooldownDurationMs(state.tierIndex, tiered)
        state.lastAlertAtMs = nowMs
        state.lastActivityAtMs = nowMs
        state.cooldownUntilMs = nowMs + durationMs
        if (tiered && state.suppressedCount > 0) {
            state.tierIndex = minOf(state.tierIndex + 1, cooldownTiersMs.lastIndex)
        }
        state.suppressedCount = 0
        stateByKey[conversationKey] = state
    }

    fun recordSilentUpdate(
        conversationKey: String,
        pushType: String = "circle.chat.new_message",
        nowMs: Long = System.currentTimeMillis(),
    ) {
        val state = stateByKey.getOrPut(conversationKey) { ConversationNotificationState() }
        applyIdleDecay(state, nowMs)
        state.suppressedCount += 1
        state.lastActivityAtMs = nowMs
        val tiered = usesTieredCooldown(pushType)
        if (state.cooldownUntilMs == null) {
            state.cooldownUntilMs = nowMs + cooldownDurationMs(state.tierIndex, tiered)
        }
        stateByKey[conversationKey] = state
    }

    fun bannerBody(base: String, suppressedCount: Int): String =
        if (suppressedCount > 0) {
            "$base (+$suppressedCount more)"
        } else {
            base
        }

    /** Clears in-memory throttle state (unit tests). */
    internal fun resetForTesting() {
        stateByKey.clear()
    }

    internal fun tierIndexForTesting(conversationKey: String): Int =
        stateByKey[conversationKey]?.tierIndex ?: 0

    internal fun cooldownUntilMsForTesting(conversationKey: String): Long? =
        stateByKey[conversationKey]?.cooldownUntilMs

    private fun usesTieredCooldown(pushType: String): Boolean =
        pushType.trim() == "circle.chat.new_message"

    private fun cooldownDurationMs(tierIndex: Int, tiered: Boolean): Long {
        if (!tiered) return cooldownTiersMs[0]
        val clamped = tierIndex.coerceIn(0, cooldownTiersMs.lastIndex)
        return cooldownTiersMs[clamped]
    }

    private fun applyIdleDecay(state: ConversationNotificationState, nowMs: Long) {
        if (state.tierIndex <= 0 || state.lastActivityAtMs == null) return
        val idleMs = nowMs - state.lastActivityAtMs!!
        val stepsDown = minOf(state.tierIndex, (idleMs / IDLE_DECAY_MS).toInt())
        if (stepsDown <= 0) return
        state.tierIndex -= stepsDown
        state.suppressedCount = 0
    }

    private fun evaluateCooldown(
        key: String,
        nowMs: Long,
    ): Pair<ChatEngagementAlertOutcome, Int> {
        val state = stateByKey.getOrPut(key) { ConversationNotificationState() }
        applyIdleDecay(state, nowMs)
        state.lastActivityAtMs = nowMs
        stateByKey[key] = state

        val cooldownUntil = state.cooldownUntilMs
        if (cooldownUntil != null && nowMs < cooldownUntil) {
            return ChatEngagementAlertOutcome.SilentBannerUpdate to (state.suppressedCount + 1)
        }
        return ChatEngagementAlertOutcome.FullAlert to 0
    }
}
